README.md
# Description

This cookbook will install RStudio Server along with some extras that will configure various usages. In addition to installing the server, it can optionally set up CRAN's distribution-specific repositories, configure various Nginx options (SSL, proxy, and alternate locations), as well as an alternative to the default PAM configuration that doe *not* rely on a local user account.

* `rstudio::server` will download and configure the RServer packages.
* `rstudio::pam` will install the PAM [pam_pwdfile](https://github.com/tiwe-de/libpam-pwdfile) module. This module allows you to define username/password combinations using the `mkpasswd` utility. This allows you to give access to RStudio Server *without* creating accounts on the server itself.
* `rstudio::cran` will install R packages listed in `node['rstudio']['cran']['packages']` using `r_package` from the [r](https://github.com/stevendanna/cookbook-r/) cookbook.
* `rstudio::nginx` will configure an Nginx site for RStudio. It offers the ability to proxy via a separate URL as well as basic SSL support. When coupled with `rstudio::pam`, you can give remote access to RStudio via Nginx over SSL to people *without* direct access to the server via SSH.
* `rstudio::shiny` will install and configure the [Shiny Server](http://www.rstudio.com/shiny/server/) packages.

# Requirements

## Cookbooks

The following Chef cookbooks are required:

* [r](https://github.com/stevendanna/cookbook-r/)
* [apt](https://github.com/opscode-cookbooks/apt)
* [nginx](https://github.com/opscode-cookbooks/nginx)
* [users](https://github.com/opscode-cookbooks/users)

## Platform

The following platforms are supported:

* Ubuntu
* Debian

By default RStudio and its related packages are installed via APT on Debian/Ubuntu. You can modify the default repository using the following attribributes.

* `node['rstudio']['apt']['uri']` - The URI of the APT repository. Defaults to `http://cran.stat.ucla.edu/bin/linux/ubuntu` on Ubuntu and `'http://cran.stat.ucla.edu/bin/linux/debian` on Debian.
* `node['rstudio']['apt']['key']` - The key for the distribution. Defaults to `E084DAB9` on Ubunto and `381BA480` on Debian.
* `node['rstudio']['apt']['keyserver']` - The key server for the repository. Defaults to `keyserver.ubuntu.com` on Ubuntu and `subkeys.pgp.net` on Debian.

# Attributes

## rstudio::server

* `node['rstudio']['server']['www_port']` - The port that RStudio Server binds to. Defaults to `8787`.
* `node['rstudio']['server']['www_address']` - The address that RStudio Server binds to. Defaults to `127.0.0.1`
* `node['rstudio']['server']['ld_library_path']` - The LD library path.
* `node['rstudio']['server']['r_binary_path']` - The path to R.
* `node['rstudio']['server']['user_group']` - User groups you'd like to restrict to. **NOTE:** The `rstudio::pam` recipe will likely ignore and/or conflict with this setting if used.
* `node['rstudio']['session']['timeout']` - The session timeout in minutes. Defaults to `30`.
* `node['rstudio']['session']['package_path']` - The R package path.
* `node['rstudio']['session']['cran_repo']` - The CRAN repository to use. Defaults to `http://cran.case.edu/`.

See also: [RStudio Server: Configuring the Server](http://www.rstudio.com/ide/docs/server/configuration)

## rstudio::pam

The `rstudio::pam` recipe works on top of the [users](https://github.com/opscode-cookbooks/users) cookbook, which uses data bags. In your various `users` data bags, put the following stanza in to create an RStudio account for them:

```
  "rstudio": {
    "passwd": "WQtdiGghp/GSU"
  }
```

You can create a password with the `mkpasswd` utility.

## rstudio::cran

* `node['rstudio']['cran']['mirror']` - The CRAN distribution repository. Defaults to `http://cran.rstudio.com/`. This value is used to set `r-cran-repos` in `rsession.conf`.
* `node['rstudio']['cran']['packages'] - A list of CRAN packages to install using the `r_package` LWRP from the [r](https://github.com/stevendanna/cookbook-r/) cookbook.

## rstudio::nginx

* `node['rstudio']['nginx']['port']` - The port to listen on. Defaults to `80`. If you specify a `node['rstudio']['ssl']['crt_file']` and `node['rstudio']['ssl']['key_file']`, this will be overridden to use `443`.
* `node['rstudio']['nginx']['server_name']` - You **must** set a server name to utilize the Nginx recipe.
* `node['rstudio']['nginx']['location']` - The location you'd like to proxy RStudio requests from. Defaults to `/`. You could change this to `/rstudio/` to change the location of RStudio.
* `node['rstudio']['ssl']['crt_file']` - An SSL certificate file to use. This enables SSL.
* `node['rstudio']['ssl']['key_file']` - Your SSL certificate's private key file.

**NOTE:** The Nginx recipe does *not* use HTTP-Auth. This is because RStudio will *only* authenticate via PAM. There is no way to turn off authentication either.

## rstudio::shiny

Installs and configures [Shiny Server](http://www.rstudio.com/shiny/server/). For more information see the documentation on [configuration settings](http://rstudio.github.io/shiny-server/latest/#configuration-settings).

* `node['rstudio']['shiny']['user']` - The user that Shiny will run as. Defaults to `shiny`.
* `node['rstudio']['shiny']['www_port']` - The port that Shiny will listen on. Defaults to `8100`.
* `node['rstudio']['shiny']['www_address']` - The address that Shiny will listen on. Defaults to `127.0.0.1`.
* `node['rstudio']['shiny']['site_dir']` - The server directory. Defaults to `/srv/shiny-server`.
* `node['rstudio']['shiny']['log_dir']` - Where the log files will reside. Defaults to `/var/log/shiny-server`.
* `node['rstudio']['shiny']['directory_index']` - Whether or not to turn on directory indexing. Defaults to `on`.
* `node['rstudio']['shiny']['htpasswd_file']` - Path to the username/password file to use for [authentication](http://nginx.org/en/docs/http/ngx_http_auth_basic_module.html). Defaults to `/etc/shiny-server/htpasswd`.
* `node['rstudio']['shiny']['htpasswd_group']` - The group to use for `node['rstudio']['shiny']['htpasswd_file']`, which requires use of the [users](https://github.com/opscode-cookbooks/users) cookbook. Defaults to `sysadmin`.

# Usage

The only thing you need to watch out for is that `rstudio::cran` is listed before `rstudio::server` if you'd like to use CRAN's packages for `r-base` and other related R packages.

```
  "run_list": [
    "recipe[rstudio::cran]",
    "recipe[rstudio::server]",
    "recipe[rstudio::nginx]",
    "recipe[rstudio::pam]"
  ]
```

This is an example configuration for a role file that sets up Nginx with SSL and modifies the default path:

```
  "default_attributes": {
    "rstudio": {
      "nginx": {
        "server_name": "data.example.com",
        "location": "/rstudio/"
      },
      "ssl": {
        "crt_file": "/path/to/example.crt",
        "key_file": "/path/to/example.key"
      }
    },
  }
```


CHANGELOG.md
rstudio CHANGELOG
=================

0.1.0
-----
- Initial release of rstudio

0.2.0
-----
- Added [Shiny Server](http://www.rstudio.com/shiny/server/) support
- Added tentative support for Debian (Joe Stump)
- Updated the APT installation to use the official APT repositories for Ubuntu and Debian
- Updated the Nginx configuration to support serving both RStudio and Shiny from the same server
- Now depends on the [r](https://github.com/stevendanna/cookbook-r/) cookbook.
- Fixed support for `node['rstudio']['cran']['packages']` using `r_package` LWRP from the `r` cookbook.

0.3.1
-----
- Test fixes and updated dependency versions

0.4.0
-----
- Shiny recipe fix via PR #10


spec/pam_spec.rb
require 'spec_helper'

describe 'rstudio::pam' do
  let(:chef_run) do
    runner = ChefSpec::SoloRunner.new platform: 'ubuntu', version: '12.04' do |node|
      node.set['rstudio']['nginx']['server_name'] = 'test.example.com'
    end.converge(described_recipe)
  end

  it('includes rstudio::server') do 
    expect(chef_run).to include_recipe('rstudio::server')
  end

  it('install libpam-pwdfile') do
    expect(chef_run).to install_package('libpam-pwdfile')
  end
end


spec/cran_spec.rb
require 'spec_helper'

describe 'rstudio::cran' do
  let(:chef_run) do
    runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '12.04')
    runner.converge(described_recipe)
  end
end


spec/nginx_spec.rb
require 'spec_helper'

describe 'rstudio::nginx' do
  let(:chef_run) do
    runner = ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '12.04')
    runner.node.set['rstudio']['nginx']['server_name'] = 'test.example.com'
    runner.converge(described_recipe)
  end

  before(:each) do 
    stub_command('which nginx').and_return(true)
  end

  it 'should create a knife.rb' do
    template_file = '/etc/nginx/sites-available/test.example.com'
    expect(chef_run).to render_file(template_file)
  end
end


recipes/nginx.rb
include_recipe "nginx"

# Only configure Nginx if the server_name attribute has been set.
if node['rstudio']['nginx']['server_name'] == ''
    raise ArgumentError, "You must specify a server_name to configure Nginx support."
else
    server_name = node['rstudio']['nginx']['server_name']
end

case node["platform"].downcase
when "ubuntu"
    template_file = "/etc/nginx/sites-available/#{server_name}"
end

if node['rstudio']['ssl']['crt_file'] != '' && node['rstudio']['ssl']['crt_file'] != ''
    use_ssl = true
else
    use_ssl = false
end

template template_file do
    source "etc/nginx/rstudio.conf.erb"
    mode 0644
    owner "root"
    group "root"
    variables({:use_ssl => use_ssl})
    notifies :reload, "service[nginx]"
end

nginx_site server_name do
  server_name server_name
  default_server true
  template "etc/nginx/rstudio.conf.erb"
end


recipes/shiny.rb
if node['rstudio']['shiny']['arch'] == 'amd64'
    base_download_url = 'http://download3.rstudio.org/ubuntu-12.04/x86_64'
else
    raise Exception, "This cookbook doesn't work with i386."
end

case node["platform"].downcase
when "ubuntu", "debian"
    include_recipe "apt"

    package "r-base"

    remote_shiny_server_file = "#{base_download_url}/shiny-server-#{node['rstudio']['shiny']['version']}-#{node['rstudio']['shiny']['arch']}.deb"
    local_shiny_server_file = "/tmp/shiny-server-shiny-server-#{node['rstudio']['shiny']['version']}-#{node['rstudio']['shiny']['arch']}.deb"
    remote_file local_shiny_server_file do
        source remote_shiny_server_file
        action :create_if_missing
        not_if { ::File.exists?('/etc/init/shiny-server.conf') }
    end

    execute "install-shiny-server" do
        command "dpkg --install #{local_shiny_server_file}"
        not_if { ::File.exists?('/etc/init/shiny-server.conf') }
    end

    r_package "shiny"
end

group "shiny"

user "shiny" do
  gid "shiny"
  action :create
end

template "/etc/init/shiny-server.conf" do
  source "etc/init/shiny-server.conf.erb"
  mode "0644"
  owner "root"
  group "root"
end

service "shiny-server" do
  provider Chef::Provider::Service::Upstart
  supports :start => true, :stop => true, :restart => true, :reload => true
  action [:enable, :start]
end

directory "/etc/shiny-server" do
  owner "root"
  group "root"
  action :create
end

directory node['rstudio']['shiny']['site_dir'] do
  owner "root"
  group "root"
  action :create
  recursive true
end

template "/etc/shiny-server/shiny-server.conf" do
  source "etc/shiny-server/shiny-server.conf.erb"
  mode "0644"
  owner "root"
  group "root"
  notifies :reload, "service[shiny-server]"
end

if node['rstudio']['shiny']['htpasswd_file'] != ''
  if Chef::Config[:solo]
    Chef::Log.warn("This recipe uses search. Chef Solo does not support search.")
  else
    users = search(:users, "groups:#{node['rstudio']['shiny']['htpasswd_group']}")

    template node['rstudio']['shiny']['htpasswd_file'] do
        source "etc/shiny-server/htpasswd.erb"
        owner "root"
        group "root"
        mode "0644"
        variables(
            :users => users
        )
        action :create
        notifies :reload, "service[nginx]"
    end
  end
end



recipes/server.rb
# Set up the package repository.
case platform_family
when "debian"
    include_recipe "apt"

    apt_repository "rstudio-cran" do
        uri node['rstudio']['apt']['uri']
        keyserver node['rstudio']['apt']['keyserver']
        key node['rstudio']['apt']['key']
        distribution "#{node['lsb']['codename']}/"
    end

    package "r-base" do
        action :install
    end

    package "libssl0.9.8" do
        action :install
    end

    Chef::Log.info('Retrieving RStudio Server file.')
    remote_rstudio_server_file = "#{node['rstudio']['server']['base_download_url']}/rstudio-server-#{node['rstudio']['server']['version']}-#{node['rstudio']['server']['arch']}.deb"
    local_rstudio_server_file = "#{Chef::Config[:file_cache_path]}/rstudio-server-#{node['rstudio']['server']['version']}-#{node['rstudio']['server']['arch']}.deb"
    remote_file local_rstudio_server_file do
        source remote_rstudio_server_file 
        action :create_if_missing
        not_if { ::File.exists?('/etc/init/shiny-server.conf') }
    end

    Chef::Log.info('Installing RStudio Server via dpkg.')
    dpkg_package "rstudio-server" do
      source local_rstudio_server_file
      action :upgrade
    end

when "rhel"
    Chef::Application.fatal!("Redhat based platforms are not yet supported")
end

service "rstudio-server" do
    provider Chef::Provider::Service::Upstart
    supports :start => true, :stop => true, :restart => true
    action :start
end

template "/etc/rstudio/rserver.conf" do
    source "etc/rstudio/rserver.conf.erb"
    mode 0644
    owner "root"
    group "root"
    notifies :restart, "service[rstudio-server]"
end

template "/etc/rstudio/rsession.conf" do
    source "etc/rstudio/rsession.conf.erb"
    mode 0644
    owner "root"
    group "root"
    notifies :restart, "service[rstudio-server]"
end


attributes/default.rb
# CRAN
default['rstudio']['cran']['mirror'] = 'https://cran.rstudio.com/'

# APT configuration for Ubuntu or Debian installs.
case node["platform"].downcase  
when "ubuntu"
    default['rstudio']['apt']['key'] = 'E084DAB9'
    default['rstudio']['apt']['keyserver'] = 'keyserver.ubuntu.com'
    default['rstudio']['apt']['uri'] = 'http://cran.stat.ucla.edu/bin/linux/ubuntu'
when "debian"
    default['rstudio']['apt']['key'] = '381BA480'
    default['rstudio']['apt']['keyserver'] = 'subkeys.pgp.net'
    default['rstudio']['apt']['uri'] = 'http://cran.stat.ucla.edu/bin/linux/debian'
end

# You can define a simple array of packages in your role/environment/node and the 
# CRAN recipe will install them.
default['rstudio']['cran']['packages'] = []

# RStudio Server
default['rstudio']['server']['base_download_url'] = 'http://download2.rstudio.org'
default['rstudio']['server']['www_port'] = '8787'
default['rstudio']['server']['www_address'] = '127.0.0.1'
default['rstudio']['server']['ld_library_path'] = ''
default['rstudio']['server']['r_binary_path'] = ''
default['rstudio']['server']['user_group'] = ''
default['rstudio']['server']['version'] = '0.99.483'
default['rstudio']['server']['arch'] = node['kernel']['machine'] =~ /x86_64/ ? "amd64" : "i386"

# RStudio Session
default['rstudio']['session']['timeout'] = '30'
default['rstudio']['session']['package_path'] = ''
default['rstudio']['session']['cran_repo'] = 'https://cran.fhcrc.org/'

# Nginx
default['rstudio']['nginx']['port'] = '80'
default['rstudio']['nginx']['server_name'] = ''
default['rstudio']['nginx']['location'] = '/'
default['rstudio']['nginx']['shiny_location'] = ''

# SSL certificates
default['rstudio']['ssl']['crt_file'] = ''
default['rstudio']['ssl']['key_file'] = ''

# Shiny server
default['rstudio']['shiny']['user'] = 'shiny'
default['rstudio']['shiny']['www_port'] = '8100'
default['rstudio']['shiny']['www_address'] = '127.0.0.1'
default['rstudio']['shiny']['site_dir'] = '/var/shiny-server/www'
default['rstudio']['shiny']['log_dir'] = '/var/log/shiny-server'
default['rstudio']['shiny']['directory_index'] = 'on'

# Shiny can't be installed by APT. Don't get me started.
default['rstudio']['shiny']['version'] = '0.4.0.8'
default['rstudio']['shiny']['arch'] = node['kernel']['machine'] =~ /x86_64/ ? "amd64" : "i386"

# Shiny server supports the users cookbook for HTTP Auth
default['rstudio']['shiny']['htpasswd_file'] = ''
default['rstudio']['shiny']['htpasswd_group'] = 'sysadmin'
